packages/@nimbus-js/runtime/src/nimbus.js

```diff
@@ -0,0 +1,276 @@
+//
+// Copyright (c) 2020, Salesforce.com, inc.
+// All rights reserved.
+// SPDX-License-Identifier: BSD-3-Clause
+// For full license text, see the LICENSE file in the repo root or
+// https://opensource.org/licenses/BSD-3-Clause
+//
+
+// @ts-check
+
+/**
+ * @typedef { import("@nimbus-js/api").NimbusPlugins } NimbusPlugins
+ */
+
+var __nimbus = (function() {
+  /**
+   * @type {NimbusPlugins}
+   */
+  let plugins = {};
+
+  // Store promise functions for later invocation
+  /**
+   *  @type { { [s: string]: { resolve: Function; reject: Function }; } }
+   */
+  let uuidsToPromises = {};
+
+  // Store callback functions for later invocation
+  //
+
+  /**
+   * @type  { { [s: string]: Function } }
+   */
+  var uuidsToCallbacks = {};
+
+  // Store event listener functions for later invocation
+  /**
+   * @type { { [s: string]: Function[] } }
+   */
+  let eventNameToListeners = {};
+
+  // influenced from
+  // https://stackoverflow.com/questions/105034/create-guid-uuid-in-javascript
+  /**
+   * @returns {string}
+   */
+  let uuidv4 = () => {
+    return "10000000-1000-4000-8000-100000000000".replace(/[018]/g, c => {
+      const asNumber = Number(c);
+      return (
+        asNumber ^
+        (crypto.getRandomValues(new Uint8Array(1))[0] & (15 >> (asNumber / 4)))
+      ).toString(16);
+    });
+  };
+
+  /**
+   * @param {any[]} args
+   * @returns {any[]}
+   */
+  let cloneArguments = args => {
+    let clonedArgs = [];
+    for (var i = 0; i < args.length; ++i) {
+      if (typeof args[i] === "function") {
+        const callbackId = uuidv4();
+        uuidsToCallbacks[callbackId] = args[i];
+        clonedArgs.push(callbackId);
+      } else if (typeof args[i] === "object") {
+        clonedArgs.push(JSON.stringify(args[i]));
+      } else {
+        clonedArgs.push(args[i]);
+      }
+    }
+    return clonedArgs;
+  };
+
+  /**
+   * @param {*} src
+   * @returns {Object}
+   */
+  let promisify = src => {
+    /** @type { {[s: string]: function(...any[]): void} } */
```

Comment:
This type should have Promise somewhere in it right?

```suggestion
    /** @type { {[s: string]: function(...any[]): Promise<any>} } */
```